# [Differential Exon-Junction Usage (DEJU)]

This is the repository for the code used to perform analysis and generate figures for the following paper titled "Incorporating exon-exon junction reads enhances differential splicing detection". The paper introduce a DEJU analysis workflow implementing a STAR-Rsubread-edgeR-limma framework to analyze DEJU and detect differential splicing between two groups of conditions. Here is a schematic presentation of our proposed DEJU workflow.

[Figure of DEJU workflow]

## **Citation**

If you are using code or pipelines from this repository, please consider citing our associated article:

Pham, M. T., Milevskiy, M. J. G., Visvader, J. E., Chen, Y. Incorporating exon-exon junction reads enhances differential splicing detection. ...

## **Code** 

The directory contains codes used for the simulation and benchmarking procedure of 6 benchmarked DEU/DEJU pipelines including DEJU-edgeR, DEU-edgeR, DEJU-limma, DEU-limma, DEXSeq, and JunctionSeq.

Overall, a standard DEU/DEJU analysis pipeline contains three main steps: Read alignment, exon-junction read quantification, and DEU/DEJU analysis.

Codes for the read alignment step can be found in the alignment folder.

Codes for the read quantification and DEU/DEJU analysis can be found in the DEU folder.

## **Reference genome**

## **Read alignment**

## **DEU/DEJU analyses**

## **Simulation**

## **Case-study**



## **Tools**

